python_modules/dagster/dagster/_core/instance/__init__.py

```diff
@@ -3139,42 +3139,46 @@ def get_latest_data_version_record(
         before_cursor: Optional[int] = None,
         after_cursor: Optional[int] = None,
     ) -> Optional["EventLogRecord"]:
-        from dagster._core.events import DagsterEventType
-        from dagster._core.storage.event_log.base import EventRecordsFilter
+        from dagster._core.storage.event_log.base import AssetRecordsFilter
 
         # When we cant don't know whether the requested key corresponds to a source or regular
         # asset, we need to retrieve both the latest observation and materialization for all assets.
         # If there is a materialization, it's a regular asset and we can ignore the observation.
-
-        observation: Optional[EventLogRecord] = None
-        if is_source or is_source is None:
-            observations = self.get_event_records(
-                EventRecordsFilter(
-                    event_type=DagsterEventType.ASSET_OBSERVATION,
-                    asset_key=key,
-                    asset_partitions=[partition_key] if partition_key else None,
-                    before_cursor=before_cursor,
-                    after_cursor=after_cursor,
+        if not is_source:
```

Comment:
I know it seems trivial, but it might be helpful to try a tiny class or helper function to interpret `is_source`, which is a cursed tri-state bool.

```python
# class for interpreting is_source
class IsSourceOracle(NamedTuple):
   is_source: Optional[bool]

   def is_materializable(self) -> bool:
       return not bool(self.is_source)

   def possible_has_observations(self) -> bool:
      if self.is_source:
        return True # explicit sources definitely have observations
      return self.is_source is None # we don't know, so it possibly might have observations
```

Then this serves as documentation for this logic, which is super non-obvious.

I think it is also worth re-examining the assumption that "regular" assets can have observations too, but that is beyond the scope of this PR.

Reply:
lol `IsSourceOracle`

Reply:
I tried writing it with some intermediate class, but it felt contrived and awkward.  I think just rewriting for clarity, with comments, is best.